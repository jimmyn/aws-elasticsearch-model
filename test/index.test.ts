import {Client} from 'elasticsearch';
import {dynamoDBStreamEvent} from './fixtures';
import {ElasticModel} from 'index';

const defaultConfig = {
  host: 'https://someurl.com',
  index: 'existing-index'
};

describe('ElasticModel', () => {
  beforeEach(() => {
    // @ts-ignore
    Client.mockClear();
  });

  it('should throw en error if required config is missing', () => {
    // @ts-ignore
    expect(() => new ElasticModel({})).toThrowErrorMatchingSnapshot();
    // @ts-ignore
    expect(() => new ElasticModel({host: 'https://someurl.com'})).toThrowErrorMatchingSnapshot();
    expect(
      () =>
        // @ts-ignore
        new ElasticModel({
          host: 'https://someurl.com'
        })
    ).toThrowErrorMatchingSnapshot();
    expect(
      () =>
        // @ts-ignore
        new ElasticModel({
          host: 'https://someurl.com',
          index: 'user'
        })
    ).not.toThrow();
  });

  test('bulkIndex should create index if missing', async () => {
    const config = {
      host: 'https://someurl.com',
      index: 'missing-index',
      settings: {
        analysis: {
          analyzer: {
            email: {
              type: 'custom',
              tokenizer: 'uax_url_email'
            }
          }
        }
      },
      mapping: {
        email: {
          type: 'email'
        },
        name: {
          type: 'text'
        }
      }
    };
    const data = [
      {
        id: 'some_id',
        email: 'example@email.com'
      }
    ];
    const model = new ElasticModel(config);
    // @ts-ignore
    const client = Client.mock.instances[0];
    const result = await model.bulkIndex(data);
    expect(client.indices.exists.mock.calls[0][0]).toMatchSnapshot();
    expect(client.indices.create.mock.calls[0][0]).toMatchSnapshot();
    return expect(result).toMatchSnapshot();
  });

  test('deleteIndex should work correctly', async () => {
    const model = new ElasticModel(defaultConfig);
    // @ts-ignore
    const client = Client.mock.instances[0];
    const result = await model.deleteIndex();
    expect(client.indices.delete).toHaveBeenCalledWith({index: defaultConfig.index});
    return expect(result).toMatchSnapshot();
  });

  test('bulkIndex should not create index if already exists', async () => {
    const data = [
      {
        id: 'some_id',
        email: 'example@email.com'
      }
    ];
    const model = new ElasticModel(defaultConfig);
    // @ts-ignore
    const client = Client.mock.instances[0];
    const result = await model.bulkIndex(data);
    expect(client.indices.exists).toHaveBeenCalledWith({index: defaultConfig.index});
    expect(client.indices.create).not.toHaveBeenCalled();
    return expect(result).toMatchSnapshot();
  });

  test('search should return correct result', () => {
    const model = new ElasticModel(defaultConfig);
    return expect(model.search({size: 5, from: 5})).resolves.toMatchSnapshot();
  });

  test('buildQuery should work correctly', async () => {
    const model = new ElasticModel(defaultConfig);
    const searchSpy = jest.spyOn(model, 'search');
    const query = model
      .buildQuery()
      .sort('createdAt', 'desc')
      .from(5)
      .size(5);

    expect(query.exec).toBeDefined();

    const result = await query.exec();
    expect(result).toMatchSnapshot();
    expect(searchSpy).toHaveBeenCalledWith({
      size: 5,
      from: 5,
      sort: [{createdAt: {order: 'desc'}}]
    });
  });

  test('indexFromDynamoDBStream should generate correct bulk request', async () => {
    const config = {
      host: 'https://someurl.com',
      index: 'existing-index'
    };
    const model = new ElasticModel(config);
    // @ts-ignore
    const client = Client.mock.instances[0];
    await model.indexFromDynamoDBStream(dynamoDBStreamEvent);
    expect(client.bulk.mock.calls[0]).toMatchSnapshot();
  });

  test('indexFromDynamoDBStream should exclude fields specified in excludedFields array', async () => {
    const config = {
      host: 'https://someurl.com',
      index: 'existing-index',
      excludedFields: ['email']
    };
    const model = new ElasticModel(config);
    // @ts-ignore
    const client = Client.mock.instances[0];
    await model.indexFromDynamoDBStream(dynamoDBStreamEvent);
    expect(client.bulk.mock.calls[0]).toMatchSnapshot();
  });

  test('bulkIndex should exclude fields specified in excludedFields array', async () => {
    const config = {
      host: 'https://someurl.com',
      index: 'existing-index',
      excludedFields: ['email']
    };
    const data = [
      {
        id: 'some_id',
        email: 'example@email.com'
      }
    ];
    const model = new ElasticModel(config);
    // @ts-ignore
    const client = Client.mock.instances[0];
    await model.bulkIndex(data);
    expect(client.bulk.mock.calls[0]).toMatchSnapshot();
  });
});
