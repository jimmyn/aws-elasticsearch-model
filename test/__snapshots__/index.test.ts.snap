// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticModel buildQuery should work correctly 1`] = `
Object {
  "aggregations": undefined,
  "items": Array [
    Object {
      "id": 0,
    },
    Object {
      "id": 1,
    },
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 4,
    },
  ],
  "raw": Object {
    "_shards": Object {
      "failed": 0,
      "skipped": 0,
      "successful": 25,
      "total": 25,
    },
    "hits": Object {
      "hits": Array [
        Object {
          "_id": 0,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 0,
          },
          "_type": undefined,
        },
        Object {
          "_id": 1,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 1,
          },
          "_type": undefined,
        },
        Object {
          "_id": 2,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 2,
          },
          "_type": undefined,
        },
        Object {
          "_id": 3,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 3,
          },
          "_type": undefined,
        },
        Object {
          "_id": 4,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 4,
          },
          "_type": undefined,
        },
      ],
      "max_score": 0,
      "total": Object {
        "relation": "eq",
        "value": 25,
      },
    },
    "timed_out": false,
    "took": 6,
  },
  "total": 25,
}
`;

exports[`ElasticModel bulkIndex should create index if missing 1`] = `
Object {
  "index": "missing-index",
}
`;

exports[`ElasticModel bulkIndex should create index if missing 2`] = `
Object {
  "body": Object {
    "mappings": Object {
      "properties": Object {
        "email": Object {
          "type": "email",
        },
        "name": Object {
          "type": "text",
        },
      },
    },
    "settings": Object {
      "analysis": Object {
        "analyzer": Object {
          "email": Object {
            "tokenizer": "uax_url_email",
            "type": "custom",
          },
        },
      },
    },
  },
  "index": "missing-index",
}
`;

exports[`ElasticModel bulkIndex should create index if missing 3`] = `
Object {
  "errors": false,
  "items": Array [
    Object {
      "index": Object {
        "_id": "some_id",
        "_index": "missing-index",
      },
    },
  ],
  "took": 6,
}
`;

exports[`ElasticModel bulkIndex should exclude fields specified in excludedFields array 1`] = `
Array [
  Object {
    "body": Array [
      Object {
        "index": Object {
          "_id": "some_id",
          "_index": "existing-index",
        },
      },
      Object {
        "id": "some_id",
      },
    ],
  },
]
`;

exports[`ElasticModel bulkIndex should not create index if already exists 1`] = `
Object {
  "errors": false,
  "items": Array [
    Object {
      "index": Object {
        "_id": "some_id",
        "_index": "existing-index",
      },
    },
  ],
  "took": 6,
}
`;

exports[`ElasticModel deleteIndex should work correctly 1`] = `
Object {
  "acknowledged": true,
}
`;

exports[`ElasticModel indexFromDynamoDBStream should exclude fields specified in excludedFields array 1`] = `
Array [
  Object {
    "body": Array [
      Object {
        "index": Object {
          "_id": "some-id",
          "_index": "existing-index",
        },
      },
      Object {
        "id": "some-id",
      },
      Object {
        "delete": Object {
          "_id": "some-id",
          "_index": "existing-index",
        },
      },
      Object {
        "update": Object {
          "_id": "some-id",
          "_index": "existing-index",
        },
      },
      Object {
        "doc": Object {
          "id": "some-id",
        },
      },
    ],
  },
]
`;

exports[`ElasticModel indexFromDynamoDBStream should generate correct bulk request 1`] = `
Array [
  Object {
    "body": Array [
      Object {
        "index": Object {
          "_id": "some-id",
          "_index": "existing-index",
        },
      },
      Object {
        "email": "email@example.com",
        "id": "some-id",
      },
      Object {
        "delete": Object {
          "_id": "some-id",
          "_index": "existing-index",
        },
      },
      Object {
        "update": Object {
          "_id": "some-id",
          "_index": "existing-index",
        },
      },
      Object {
        "doc": Object {
          "email": "email@example.com",
          "id": "some-id",
        },
      },
    ],
  },
]
`;

exports[`ElasticModel search should return correct result 1`] = `
Object {
  "aggregations": undefined,
  "items": Array [
    Object {
      "id": 0,
    },
    Object {
      "id": 1,
    },
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 4,
    },
  ],
  "raw": Object {
    "_shards": Object {
      "failed": 0,
      "skipped": 0,
      "successful": 25,
      "total": 25,
    },
    "hits": Object {
      "hits": Array [
        Object {
          "_id": 0,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 0,
          },
          "_type": undefined,
        },
        Object {
          "_id": 1,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 1,
          },
          "_type": undefined,
        },
        Object {
          "_id": 2,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 2,
          },
          "_type": undefined,
        },
        Object {
          "_id": 3,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 3,
          },
          "_type": undefined,
        },
        Object {
          "_id": 4,
          "_index": "existing-index",
          "_score": 0,
          "_source": Object {
            "id": 4,
          },
          "_type": undefined,
        },
      ],
      "max_score": 0,
      "total": Object {
        "relation": "eq",
        "value": 25,
      },
    },
    "timed_out": false,
    "took": 6,
  },
  "total": 25,
}
`;

exports[`ElasticModel should throw en error if required config is missing 1`] = `"Required config \\"host\\" is missing!"`;

exports[`ElasticModel should throw en error if required config is missing 2`] = `"Required config \\"index\\" is missing!"`;

exports[`ElasticModel should throw en error if required config is missing 3`] = `"Required config \\"index\\" is missing!"`;
